File: services/mappls.service.js
import axios from "axios";

// Token cache
let cachedToken = null;
let tokenExpiry = null;

/**
 * Get OAuth2 access token from Mappls API
 * Token is cached for 23 hours (expires in 24h)
 */
export async function getAccessToken() {
    // Return cached token if still valid
    if (cachedToken && tokenExpiry && Date.now() < tokenExpiry) {
        return cachedToken;
    }

    try {
        const response = await axios.post(
            "https://outpost.mappls.com/api/security/oauth/token",
            new URLSearchParams({
                grant_type: "client_credentials",
                client_id: process.env.MAPPLS_CLIENT_ID,
                client_secret: process.env.MAPPLS_CLIENT_SECRET
            }),
            {
                headers: {
                    "Content-Type": "application/x-www-form-urlencoded"
                }
            }
        );

        cachedToken = response.data.access_token;
        // Cache for 23 hours (token expires in 24h)
        tokenExpiry = Date.now() + (23 * 60 * 60 * 1000);

        console.log("Mappls token generated successfully");
        return cachedToken;
    } catch (error) {
        console.error("Error getting Mappls access token:", error.response?.data || error.message);
        throw new Error("Failed to authenticate with Mappls");
    }
}

/**
 * Geocode a location name to coordinates using Geoapify API
 * (Mappls geocode doesn't return lat/lng directly, so we use Geoapify)
 */
export async function geocode(location) {
    try {
        const GEOAPIFY_API_KEY = process.env.PLACES_API_KEY;
        const response = await axios.get(
            `https://api.geoapify.com/v1/geocode/search`,
            {
                params: {
                    text: location,
                    apiKey: GEOAPIFY_API_KEY,
                    limit: 1
                }
            }
        );

        if (response.data.features && response.data.features.length > 0) {
            const feature = response.data.features[0];
            return {
                lat: feature.properties.lat,
                lng: feature.properties.lon,
                name: feature.properties.formatted || location
            };
        }

        throw new Error("Geoapify returned no results");
    } catch (error) {
        console.log(`Geoapify failed for ${location}, trying Nominatim fallback...`);
        try {
            // Fallback to Nominatim (OpenStreetMap)
            const response = await axios.get(
                `https://nominatim.openstreetmap.org/search`,
                {
                    params: {
                        q: location,
                        format: 'json',
                        limit: 1
                    },
                    headers: {
                        'User-Agent': 'TripPlannerApp/1.0'
                    }
                }
            );

            if (response.data && response.data.length > 0) {
                const feature = response.data[0];
                return {
                    lat: parseFloat(feature.lat),
                    lng: parseFloat(feature.lon),
                    name: feature.display_name || location
                };
            }
        } catch (nominatimError) {
            console.error("Nominatim fallback also failed:", nominatimError.message);
        }

        console.error("Geocoding failed for:", location, error.message);
        return null; // Both failed
    }
}

/**
 * Get route between source and destination
 * Returns polyline coordinates for map display
 */
export async function getRoute(sourceCoords, destCoords) {
    try {
        const token = await getAccessToken();

        // Mappls Directions API
        const response = await axios.get(
            `https://apis.mappls.com/advancedmaps/v1/${token}/route_adv/driving/${sourceCoords.lng},${sourceCoords.lat};${destCoords.lng},${destCoords.lat}`,
            {
                params: {
                    geometries: "polyline",
                    overview: "full",
                    steps: true
                }
            }
        );

        if (response.data.routes && response.data.routes.length > 0) {
            const route = response.data.routes[0];
            return {
                distance: route.distance, // in meters
                duration: route.duration, // in seconds
                geometry: route.geometry, // encoded polyline
                steps: route.legs?.[0]?.steps || []
            };
        }

        return null;
    } catch (error) {
        console.error("Error getting route:", error.response?.data || error.message);
        return null;
    }
}

/**
 * Get formatted distance and duration between two locations
 */
export async function getDistanceInfo(source, destination) {
    try {
        const [sourceCoords, destCoords] = await Promise.all([
            geocode(source),
            geocode(destination)
        ]);

        if (!sourceCoords || !destCoords) {
            return null;
        }

        const route = await getRoute(sourceCoords, destCoords);

        if (!route) {
            return null;
        }

        // Format distance (meters to km)
        const distanceKm = (route.distance / 1000).toFixed(1);

        // Format duration (seconds to hours and minutes)
        const hours = Math.floor(route.duration / 3600);
        const minutes = Math.floor((route.duration % 3600) / 60);
        const durationStr = hours > 0
            ? `${hours} hour${hours > 1 ? 's' : ''} ${minutes} min`
            : `${minutes} minutes`;

        return {
            distanceKm: parseFloat(distanceKm),
            distanceText: `${distanceKm} km`,
            durationSeconds: route.duration,
            durationText: durationStr,
            source: sourceCoords,
            destination: destCoords
        };
    } catch (error) {
        console.error("Error getting distance info:", error);
        return null;
    }
}

/**
 * Search for places/attractions near a location
 */
export async function searchPlaces(lat, lng, query = "tourist attraction", radius = 10000) {
    try {
        const token = await getAccessToken();

        const response = await axios.get(
            `https://atlas.mappls.com/api/places/nearby/json`,
            {
                params: {
                    keywords: query,
                    refLocation: `${lat},${lng}`,
                    radius: radius,
                    page: 1
                },
                headers: {
                    Authorization: `Bearer ${token}`
                }
            }
        );

        if (response.data.suggestedLocations) {
            return response.data.suggestedLocations.map(place => ({
                name: place.placeName,
                address: place.placeAddress,
                lat: parseFloat(place.latitude),
                lng: parseFloat(place.longitude),
                type: place.type,
                eLoc: place.eLoc
            }));
        }

        return [];
    } catch (error) {
        console.error("Error searching places:", error.response?.data || error.message);
        return [];
    }
}

/**
 * Get complete map data for a trip
 */
export async function getMapData(source, destination) {
    try {
        // Geocode source and destination
        const [sourceCoords, destCoords] = await Promise.all([
            geocode(source),
            geocode(destination)
        ]);

        if (!sourceCoords || !destCoords) {
            return {
                error: "Could not geocode source or destination",
                source: sourceCoords,
                destination: destCoords
            };
        }

        // Get route and nearby places in parallel
        const [route, places] = await Promise.all([
            getRoute(sourceCoords, destCoords),
            searchPlaces(destCoords.lat, destCoords.lng)
        ]);

        return {
            source: sourceCoords,
            destination: destCoords,
            route: route,
            places: places.slice(0, 10) // Limit to 10 places
        };
    } catch (error) {
        console.error("Error getting map data:", error);
        return { error: error.message };
    }
}

File: services/mappls_service.py
import os
import time
import requests
import logging

logger = logging.getLogger(__name__)

# Token cache
_cached_token = None
_token_expiry = 0

def get_access_token():
    global _cached_token, _token_expiry
    
    if _cached_token and _token_expiry and time.time() < _token_expiry:
        return _cached_token

    try:
        url = "https://outpost.mappls.com/api/security/oauth/token"
        data = {
            "grant_type": "client_credentials",
            "client_id": os.getenv("MAPPLS_CLIENT_ID"),
            "client_secret": os.getenv("MAPPLS_CLIENT_SECRET")
        }
        headers = {"Content-Type": "application/x-www-form-urlencoded"}
        
        response = requests.post(url, data=data, headers=headers)
        response.raise_for_status()
        
        token_data = response.json()
        _cached_token = token_data.get("access_token")
        # Cache for 23 hours
        _token_expiry = time.time() + (23 * 60 * 60)
        
        logger.info("Mappls token generated successfully")
        return _cached_token
        
    except Exception as e:
        logger.error(f"Error getting Mappls access token: {e}")
        return None

def geocode(location):
    try:
        geoapify_key = os.getenv("GEOAPIFY_API_KEY") or os.getenv("PLACES_API_KEY")
        
        if not geoapify_key:
            logger.error("❌ GEOAPIFY_API_KEY not configured")
            return None

        # Try Geoapify
        url = "https://api.geoapify.com/v1/geocode/search"
        params = {"text": location, "apiKey": geoapify_key, "limit": 1}
        
        response = requests.get(url, params=params)
        
        if response.status_code == 200:
            data = response.json()
            if data.get("features"):
                feat = data["features"][0]
                return {
                    "lat": feat["properties"]["lat"],
                    "lng": feat["properties"]["lon"],
                    "name": feat["properties"]["formatted"] or location
                }

        logger.warning(f"Geoapify failed for {location}, trying Nominatim...")
        
        # Fallback Nominatim
        nom_url = "https://nominatim.openstreetmap.org/search"
        nom_params = {
            "q": location, 
            "format": "json", 
            "limit": 1
        }
        nom_headers = {"User-Agent": "TripPlannerApp/1.0"}
        
        nom_res = requests.get(nom_url, params=nom_params, headers=nom_headers)
        if nom_res.status_code == 200:
            data = nom_res.json()
            if data:
                feat = data[0]
                return {
                    "lat": float(feat["lat"]),
                    "lng": float(feat["lon"]),
                    "name": feat.get("display_name") or location
                }

        return None

    except Exception as e:
        logger.error(f"Geocoding failed for {location}: {e}")
        return None

def get_route(source_coords, dest_coords):
    try:
        token = get_access_token()
        if not token: 
            return None

        url = f"https://apis.mappls.com/advancedmaps/v1/{token}/route_adv/driving/{source_coords['lng']},{source_coords['lat']};{dest_coords['lng']},{dest_coords['lat']}"
        params = {
            "geometries": "polyline",
            "overview": "full",
            "steps": "true"
        }
        
        response = requests.get(url, params=params)
        if response.status_code == 200:
            data = response.json()
            if data.get("routes"):
                route = data["routes"][0]
                return {
                    "distance": route.get("distance"),
                    "duration": route.get("duration"),
                    "geometry": route.get("geometry"),
                    "steps": route.get("legs", [{}])[0].get("steps", [])
                }
        return None
    except Exception as e:
        logger.error(f"Error getting route: {e}")
        return None

def get_distance_info(source, destination):
    try:
        # Resolve coordinates
        source_coords = geocode(source)
        dest_coords = geocode(destination) # Parallel not easily doable here without threads/async, doing sequential for simplicity

        if not source_coords or not dest_coords:
            return None

        route = get_route(source_coords, dest_coords)
        if not route:
            return None

        distance_km = round(route["distance"] / 1000, 1)
        
        total_seconds = route["duration"]
        hours = int(total_seconds // 3600)
        minutes = int((total_seconds % 3600) // 60)
        
        duration_text = f"{hours} hour{'s' if hours != 1 else ''} {minutes} min" if hours > 0 else f"{minutes} minutes"

        return {
            "distanceKm": distance_km,
            "distanceText": f"{distance_km} km",
            "durationSeconds": route["duration"],
            "durationText": duration_text,
            "source": source_coords,
            "destination": dest_coords
        }
    except Exception as e:
        logger.error(f"Error getting distance info: {e}")
        return None

def search_places(lat, lng, query="tourist attraction", radius=5000):
    try:
        geoapify_key = os.getenv("PLACES_API_KEY") or os.getenv("GEOAPIFY_API_KEY")
        
        url = "https://api.geoapify.com/v2/places"
        params = {
            "categories": "tourism.sights,tourism.attraction,entertainment.museum,religion,natural",
            "filter": f"circle:{lng},{lat},{radius}",
            "bias": f"proximity:{lng},{lat}",
            "limit": 15,
            "apiKey": geoapify_key
        }
        
        response = requests.get(url, params=params)
        if response.status_code == 200:
            data = response.json()
            results = []
            for feat in data.get("features", []):
                props = feat["properties"]
                name = props.get("name") or props.get("formatted")
                if name:
                    results.append({
                        "name": name,
                        "address": props.get("address_line2") or props.get("formatted"),
                        "lat": props.get("lat"),
                        "lng": props.get("lon"),
                        "type": props.get("categories", ["point_of_interest"])[0],
                        "eLoc": None
                    })
            return results
        return []
    except Exception as e:
        logger.error(f"Error searching places (Geoapify): {e}")
        return []

def get_map_data(source, destination):
    try:
        source_coords = geocode(source)
        dest_coords = geocode(destination)

        if not source_coords or not dest_coords:
            return {
                "error": "Could not geocode source or destination",
                "source": source_coords,
                "destination": dest_coords
            }

        route = get_route(source_coords, dest_coords)
        places = search_places(dest_coords['lat'], dest_coords['lng'])

        return {
            "source": source_coords,
            "destination": dest_coords,
            "route": route,
            "places": places[:10]
        }
    except Exception as e:
        logger.error(f"Error getting map data: {e}")
        return {"error": str(e)}

# Aliases for compatibility
get_coordinates = geocode

def get_route_data(source_coords, dest_coords, access_token=None):
    return get_route(source_coords, dest_coords)
